lib/resty/etcd/v3.lua

```diff
@@ -67,30 +94,68 @@ local function _request_uri(self, endpoint, method, uri, opts, timeout, ignore_a
         http_cli:set_timeout(timeout * 1000)
     end
 
-    local res
-    res, err = http_cli:request_uri(uri, {
-        method = method,
-        body = body,
-        headers = headers,
-        keepalive = keepalive,
-        ssl_verify = self.ssl_verify,
-        ssl_cert_path = self.ssl_cert_path,
-        ssl_key_path = self.ssl_key_path,
-    })
+    local function http_request()
+        local endpoint
+        endpoint, err = choose_endpoint(self)
+        if not endpoint then
+            return nil, err
+        end
 
-    if err then
-        if health_check.conf ~= nil then
-            health_check.report_failure(endpoint.http_host)
-            err = endpoint.http_host .. ": " .. err
+        local full_uri
+        if utils.starts_with(uri, "/version") or utils.starts_with(uri, "/v2") then
+            full_uri = endpoint.http_host .. uri
+        else
+            full_uri = endpoint.full_prefix .. uri
         end
-        return nil, err
+
+        local res
+        res, err = http_cli:request_uri(full_uri, {
+            method = method,
+            body = body,
+            headers = headers,
+            keepalive = keepalive,
+            ssl_verify = self.ssl_verify,
+            ssl_cert_path = self.ssl_cert_path,
+            ssl_key_path = self.ssl_key_path,
+        })
+
+        if err then
+            if health_check.conf ~= nil then
+                health_check.report_failure(endpoint.http_host)
+                err = endpoint.http_host .. ": " .. err
+            end
+            return nil, err
+        end
+
+        if res.status >= 500 then
+            if health_check.conf ~= nil then
+                health_check.report_failure(endpoint.http_host)
+            end
+            return nil, "invalid response code: " .. res.status
+        end
+
+        return res
     end
 
-    if res.status >= 500 then
-        if health_check.conf ~= nil then
-            health_check.report_failure(endpoint.http_host)
+    local res
+    if health_check.conf == nil or not health_check.conf.retry then
+        res, err = http_request()
+        if err then
+            return nil, err
+        end
+    else
+        local max_retry = #self.endpoints * health_check.conf.max_fails + 1
```

Comment:
Should be `#self.endpoints * health_check.conf.max_fails - 1` as we already took one try.

Reply:
Thanks for pointing it out! Fixed